### 1. Train a Fashion MNIST Classifier using SGD — Practice

## Part 1 — Implementation

In [1]:
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense
from tensorflow.keras.optimizers import SGD
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0
model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])
model.compile(
    optimizer=SGD(),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
history = model.fit(
    x_train,
    y_train,
    epochs=5,
    validation_split=0.18
)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 9s 0us/step 
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 1us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


C:\Users\lenovo\AppData\Roaming\Python\Python313\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.7521 - loss: 0.7776 - val_accuracy: 0.8076 - val_loss: 0.5739
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.8186 - loss: 0.5373 - val_accuracy: 0.8217 - val_loss: 0.5117
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8327 - loss: 0.4883 - val_accuracy: 0.8343 - val_loss: 0.4744
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8412 - loss: 0.4611 - val_accuracy: 0.8273 - val_loss: 0.4773
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.8470 - loss: 0.4412 - val_accuracy: 0.8399 - val_loss: 0.4467


### 2. Replace SGD with RMSProp

## Part 2 — Implementation

In [2]:
from tensorflow.keras.optimizers import RMSprop
model.compile(
    optimizer=RMSprop(),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
history = model.fit(
    x_train,
    y_train,
    epochs=5,
    validation_split=0.18
)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.8331 - loss: 0.4706 - val_accuracy: 0.8481 - val_loss: 0.4124
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.8606 - loss: 0.3839 - val_accuracy: 0.8603 - val_loss: 0.3925
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8736 - loss: 0.3496 - val_accuracy: 0.8720 - val_loss: 0.3766
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.8821 - loss: 0.3295 - val_accuracy: 0.8787 - val_loss: 0.3503
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.8866 - loss: 0.3192 - val_accuracy: 0.8758 - val_loss: 0.3581


### 3. Adam Optimizer with Learning Rate = 0.0005

## Part 3 — Implementation

In [3]:
from tensorflow.keras.optimizers import Adam
model.compile(
    optimizer=Adam(step_size=0.0005),
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
history = model.fit(
    x_train,
    y_train,
    epochs=5,
    validation_split=0.18
)

Epoch 1/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9019 - loss: 0.2718 - val_accuracy: 0.8864 - val_loss: 0.3194
Epoch 2/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.9054 - loss: 0.2544 - val_accuracy: 0.8856 - val_loss: 0.3163
Epoch 3/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9097 - loss: 0.2448 - val_accuracy: 0.8872 - val_loss: 0.3154
Epoch 4/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9121 - loss: 0.2355 - val_accuracy: 0.8862 - val_loss: 0.3152
Epoch 5/5
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.9172 - loss: 0.2260 - val_accuracy: 0.8914 - val_loss: 0.3063


### 4. Learning Rate Scheduler with Momentum Optimizer

## Part 4 — Implementation

In [4]:
from tensorflow.keras.optimizers import SGD
from tensorflow.keras.callbacks import LearningRateScheduler
initial_lr = 0.01
def scheduler(epoch, lr):
    if epoch > 0 and epoch % 2 == 0:
        return lr * 0.47
    return lr
callback = LearningRateScheduler(scheduler)
optimizer = SGD(
    step_size=initial_lr,
    momentum=0.87
)
model.compile(
    optimizer=optimizer,
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)
history = model.fit(
    x_train,
    y_train,
    epochs=6,
    validation_split=0.18,
    callbacks=[callback]
)

Epoch 1/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9145 - loss: 0.2267 - val_accuracy: 0.8873 - val_loss: 0.3217 - learning_rate: 0.0100
Epoch 2/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9192 - loss: 0.2189 - val_accuracy: 0.8901 - val_loss: 0.3087 - learning_rate: 0.0100
Epoch 3/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9286 - loss: 0.1926 - val_accuracy: 0.8905 - val_loss: 0.3118 - learning_rate: 0.0050
Epoch 4/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.9307 - loss: 0.1881 - val_accuracy: 0.8922 - val_loss: 0.3158 - learning_rate: 0.0050
Epoch 5/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9359 - loss: 0.1746 - val_accuracy: 0.8968 - val_loss: 0.3002 - learning_rate: 0.0025
Epoch 6/6
1500/1500 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.9380 - loss: 0.1723 - val_accuracy: 0.8954 - val_loss: 0.3051 - learning_rate: 0.0025


# 5. Why Adam is Often Preferred

Adam is one of the most popular optimizers because it combines the advantages of **Momentum** and **RMSProp**. Instead of using the same learning rate for every parameter, Adam automatically adjusts the learning rate for each weight based on past gradients. This usually leads to **faster convergence**, **less manual tuning**, and **better performance** on many deep learning problems. Because of these properties, Adam is widely used for image classification, natural language processing (NLP), and recommendation systems.

## Real-world Example

Consider **Spotify's music recommendation system**. The recommendation model is trained on millions of user interactions. If the optimizer converges slowly (such as plain SGD), training can take much longer and may produce less accurate recommendations. Using Adam allows the model to learn user preferences more efficiently, resulting in more relevant song recommendations and a better listening experience for users.

## Overall Comparison

| Optimizer | Learning Rate | Validation Accuracy (Typical) | Training Speed |
|-----------|---------------|-------------------------------|----------------|
| SGD | Default | ~84% | Slow |
| RMSProp | Default | ~88% | Medium |
| Adam | 0.0005 | ~89–90% | Fast |
| SGD + Momentum + LR Scheduler | Starts at 0.01 | ~88–89% | Stable |

## Conclusion

For this simple Fashion MNIST classifier, **Adam** with a learning rate of **0.0005** typically provides the best balance of training speed and accuracy. **RMSProp** generally outperforms plain **SGD**, while **SGD combined with Momentum and a learning-rate scheduler** offers a strong alternative with stable convergence.

## Final Notes

Review the outputs after running each section. Try changing one input value or parameter and observe how the result changes.